pkg/kubernetes/upgrade.go

```diff
@@ -0,0 +1,137 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+package kubernetes
+
+import (
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/dapr/cli/pkg/print"
+	"github.com/dapr/cli/utils"
+	helm "helm.sh/helm/v3/pkg/action"
+	"k8s.io/helm/pkg/strvals"
+)
+
+var crds = []string{
+	"components",
+	"configuration",
+	"subscription",
+}
+
+type UpgradeConfig struct {
+	RuntimeVersion string
+}
+
+func Upgrade(conf UpgradeConfig) error {
+	sc, err := NewStatusClient()
+	if err != nil {
+		return err
+	}
+
+	status, err := sc.Status()
+	if err != nil {
+		return err
+	}
+
+	if len(status) == 0 {
+		return errors.New("dapr is not installed in your cluster")
+	}
+
+	print.InfoStatusEvent(os.Stdout, "Dapr control plane version %s detected", status[0].Version)
+
+	helmConf, err := helmConfig(status[0].Namespace)
+	if err != nil {
+		return err
+	}
+
+	daprChart, err := daprChart(conf.RuntimeVersion, helmConf)
+	if err != nil {
+		return err
+	}
+
+	upgradeClient := helm.NewUpgrade(helmConf)
+	upgradeClient.ResetValues = true
+	upgradeClient.Namespace = status[0].Namespace
+	upgradeClient.CleanupOnFail = true
+	upgradeClient.Wait = true
+
+	print.InfoStatusEvent(os.Stdout, "Starting upgrade...")
+
+	mtls, err := IsMTLSEnabled()
+	if err != nil {
+		return err
+	}
+
+	var vals map[string]interface{}
+
+	if mtls {
+		secret, sErr := getTrustChainSecret()
+		if sErr != nil {
+			return sErr
+		}
+
+		ca := secret.Data["ca.crt"]
+		issuerCert := secret.Data["issuer.crt"]
+		issuerKey := secret.Data["issuer.key"]
+
+		ha := highAvailabilityEnabled(status)
```

Comment:
Can't ha be independent of mttls? So shouldn't ha values in the chart be set independently?

Reply:
Changed